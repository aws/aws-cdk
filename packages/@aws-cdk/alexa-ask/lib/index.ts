// Alexa::ASK CloudFormation Resources:
export * from './ask.generated';
export * from './pipeline-actions';
