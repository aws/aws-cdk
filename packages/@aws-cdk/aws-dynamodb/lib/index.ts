export * from './dynamodb.generated';
export * from './table';
