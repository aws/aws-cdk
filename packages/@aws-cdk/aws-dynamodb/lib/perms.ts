export const READ_DATA_ACTIONS = [
  'dynamodb:BatchGetItem',
  'dynamodb:GetRecords',
  'dynamodb:GetShardIterator',
  'dynamodb:Query',
  'dynamodb:GetItem',
  'dynamodb:Scan',
  'dynamodb:ConditionCheckItem',
];
export const KEY_READ_ACTIONS = [
  'kms:Decrypt',
  'kms:DescribeKey',
];

export const WRITE_DATA_ACTIONS = [
  'dynamodb:BatchWriteItem',
  'dynamodb:PutItem',
  'dynamodb:UpdateItem',
  'dynamodb:DeleteItem',
];
export const KEY_WRITE_ACTIONS = [
  'kms:Encrypt',
  'kms:ReEncrypt*',
  'kms:GenerateDataKey*',
];

export const READ_STREAM_DATA_ACTIONS = [
  'dynamodb:DescribeStream',
  'dynamodb:GetRecords',
  'dynamodb:GetShardIterator',
];
