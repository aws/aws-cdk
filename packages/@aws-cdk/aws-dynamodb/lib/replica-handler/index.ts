/* eslint-disable no-console */
import type { IsCompleteRequest, IsCompleteResponse, OnEventRequest, OnEventResponse } from '@aws-cdk/custom-resources/lib/provider-framework/types';
import { DynamoDB } from 'aws-sdk'; // eslint-disable-line import/no-extraneous-dependencies

export async function onEventHandler(event: OnEventRequest): Promise<OnEventResponse> {
  console.log('Event: %j', event);

  /**
   * Process only Create and Delete requests. We shouldn't receive any
   * update request and in case we do there is nothing to update.
   */
  if (event.RequestType === 'Create' || event.RequestType === 'Delete') {
    const dynamodb = new DynamoDB();

    const data = await dynamodb.updateTable({
      TableName: event.ResourceProperties.TableName,
      ReplicaUpdates: [
        {
          [event.RequestType]: {
            RegionName: event.ResourceProperties.Region,
          },
        },
      ],
    }).promise();
    console.log('Update table: %j', data);
  }

  return { PhysicalResourceId: event.ResourceProperties.Region };
}

export async function isCompleteHandler(event: IsCompleteRequest): Promise<IsCompleteResponse> {
  console.log('Event: %j', event);

  const dynamodb = new DynamoDB();

  const data = await dynamodb.describeTable({
    TableName: event.ResourceProperties.TableName,
  }).promise();
  console.log('Describe table: %j', data);

  const tableActive = !!(data.Table?.TableStatus === 'ACTIVE');
  const replicas = data.Table?.Replicas ?? [];
  const regionReplica = replicas.find(r => r.RegionName === event.ResourceProperties.Region);
  const replicaActive = !!(regionReplica?.ReplicaStatus === 'ACTIVE');

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      // Complete when replica is reported as ACTIVE
      return { IsComplete: tableActive && replicaActive };
    case 'Delete':
      // Complete when replica is gone
      return { IsComplete: tableActive && regionReplica === undefined };
  }
}
