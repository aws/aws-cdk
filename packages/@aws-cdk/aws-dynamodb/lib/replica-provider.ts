import * as path from 'path';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import { Duration, NestedStack, Stack } from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';
import { Construct } from 'constructs';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct as CoreConstruct } from '@aws-cdk/core';

/**
 * Properties for a ReplicaProvider
 */
export interface ReplicaProviderProps {
  /**
   * The timeout for the replication operation.
   *
   * @default Duration.minutes(30)
   */
  readonly timeout?: Duration;
}

export class ReplicaProvider extends NestedStack {
  /**
   * Creates a stack-singleton resource provider nested stack.
   */
  public static getOrCreate(scope: Construct, props: ReplicaProviderProps = {}) {
    const stack = Stack.of(scope);
    const uid = '@aws-cdk/aws-dynamodb.ReplicaProvider';
    return stack.node.tryFindChild(uid) as ReplicaProvider ?? new ReplicaProvider(stack, uid, props);
  }

  /**
   * The custom resource provider.
   */
  public readonly provider: cr.Provider;

  /**
   * The onEvent handler
   */
  public readonly onEventHandler: lambda.Function;

  /**
   * The isComplete handler
   */
  public readonly isCompleteHandler: lambda.Function;

  private constructor(scope: Construct, id: string, props: ReplicaProviderProps = {}) {
    super(scope as CoreConstruct, id);

    const code = lambda.Code.fromAsset(path.join(__dirname, 'replica-handler'));

    // Issues UpdateTable API calls
    this.onEventHandler = new lambda.Function(this, 'OnEventHandler', {
      code,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.onEventHandler',
      timeout: Duration.minutes(5),
    });

    // Checks if table is back to `ACTIVE` state
    this.isCompleteHandler = new lambda.Function(this, 'IsCompleteHandler', {
      code,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.isCompleteHandler',
      timeout: Duration.seconds(30),
    });

    // Allows the creation of the `AWSServiceRoleForDynamoDBReplication` service linked role
    this.onEventHandler.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['iam:CreateServiceLinkedRole'],
        resources: [Stack.of(this).formatArn({
          service: 'iam',
          region: '', // IAM is region-less
          resource: 'role',
          resourceName: 'aws-service-role/replication.dynamodb.amazonaws.com/AWSServiceRoleForDynamoDBReplication',
        })],
      }),
    );

    // Required for replica table creation
    this.onEventHandler.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ['dynamodb:DescribeLimits'],
        resources: ['*'],
      }),
    );

    this.provider = new cr.Provider(this, 'Provider', {
      onEventHandler: this.onEventHandler,
      isCompleteHandler: this.isCompleteHandler,
      queryInterval: Duration.seconds(10),
      totalTimeout: props.timeout,
    });
  }
}
