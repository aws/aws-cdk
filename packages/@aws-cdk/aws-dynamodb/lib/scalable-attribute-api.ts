import * as appscaling from '@aws-cdk/aws-applicationautoscaling';

/**
 * Interface for scalable attributes
 */
export interface IScalableTableAttribute {
  /**
   * Add scheduled scaling for this scaling attribute
   */
  scaleOnSchedule(id: string, actions: appscaling.ScalingSchedule): void;

  /**
   * Scale out or in to keep utilization at a given level
   */
  scaleOnUtilization(props: UtilizationScalingProps): void;
}

/**
 * Properties for enabling DynamoDB capacity scaling
 */
export interface EnableScalingProps {
  /**
   * Minimum capacity to scale to
   */
  readonly minCapacity: number;

  /**
   * Maximum capacity to scale to
   */
  readonly maxCapacity: number;
}

/**
 * Properties for enabling DynamoDB utilization tracking
 */
export interface UtilizationScalingProps extends appscaling.BaseTargetTrackingProps {
  /**
   * Target utilization percentage for the attribute
   */
  readonly targetUtilizationPercent: number;
}
