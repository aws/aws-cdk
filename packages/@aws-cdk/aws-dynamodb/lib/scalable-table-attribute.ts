import * as appscaling from '@aws-cdk/aws-applicationautoscaling';
import { UtilizationScalingProps } from './scalable-attribute-api';

/**
 * A scalable table attribute
 */
export class ScalableTableAttribute extends appscaling.BaseScalableAttribute {
  /**
   * Scale out or in based on time
   */
  public scaleOnSchedule(id: string, action: appscaling.ScalingSchedule) {
    super.doScaleOnSchedule(id, action);
  }

  /**
   * Scale out or in to keep utilization at a given level
   */
  public scaleOnUtilization(props: UtilizationScalingProps) {
    if (props.targetUtilizationPercent < 10 || props.targetUtilizationPercent > 90) {
      // tslint:disable-next-line:max-line-length
      throw new RangeError(`targetUtilizationPercent for DynamoDB scaling must be between 10 and 90 percent, got: ${props.targetUtilizationPercent}`);
    }
    const predefinedMetric = this.props.dimension.indexOf('ReadCapacity') === -1
      ? appscaling.PredefinedMetric.DYANMODB_WRITE_CAPACITY_UTILIZATION
      : appscaling.PredefinedMetric.DYNAMODB_READ_CAPACITY_UTILIZATION;

    super.doScaleToTrackMetric('Tracking', {
      policyName: props.policyName,
      disableScaleIn: props.disableScaleIn,
      scaleInCooldown: props.scaleInCooldown,
      scaleOutCooldown: props.scaleOutCooldown,
      targetValue: props.targetUtilizationPercent,
      predefinedMetric,
    });
  }
}

/**
 * Properties for enabling DynamoDB capacity scaling
 */
export interface EnableScalingProps {
  /**
   * Minimum capacity to scale to
   */
  minCapacity: number;

  /**
   * Maximum capacity to scale to
   */
  maxCapacity: number;
}
