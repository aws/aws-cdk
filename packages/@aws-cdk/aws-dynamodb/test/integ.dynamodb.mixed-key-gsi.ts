/**
 * This aimes to verify we can deploy a DynamoDB table with an attribute being
 * a key attribute in one GSI, and a non-key attribute in another.
 *
 * See https://github.com/aws/aws-cdk/issues/4398
 */

import { App, RemovalPolicy, Stack } from '@aws-cdk/core';
import { AttributeType, ProjectionType, Table } from '../lib';

const app = new App();
const stack = new Stack(app, 'aws-cdk-dynamodb-gsi');

const table = new Table(stack, 'Table', {
  partitionKey: { name: 'pkey', type: AttributeType.NUMBER },
  removalPolicy: RemovalPolicy.DESTROY,
});

table.addGlobalSecondaryIndex({
  indexName: 'IndexA',
  partitionKey: { name: 'foo', type: AttributeType.STRING },
  projectionType: ProjectionType.INCLUDE,
  nonKeyAttributes: ['bar'],
});

table.addGlobalSecondaryIndex({
  indexName: 'IndexB',
  partitionKey: { name: 'baz', type: AttributeType.STRING },
  sortKey: { name: 'bar', type: AttributeType.STRING },
  projectionType: ProjectionType.INCLUDE,
  nonKeyAttributes: ['blah'],
});
