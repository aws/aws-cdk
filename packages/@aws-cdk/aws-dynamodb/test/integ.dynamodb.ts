import { App, Stack } from '@aws-cdk/cdk';
import { Attribute, AttributeType, ProjectionType, StreamViewType, Table } from '../lib';

// CDK parameters
const STACK_NAME = 'aws-cdk-dynamodb';

// DynamoDB table parameters
const TABLE_PARTITION_KEY: Attribute = { name: 'hashKey', type: AttributeType.String };
const TABLE_SORT_KEY: Attribute = { name: 'sortKey', type: AttributeType.Number };

// DynamoDB global secondary index parameters
const GSI_TEST_CASE_1 = 'PartitionKeyOnly';
const GSI_TEST_CASE_2 = 'PartitionAndSortKeyWithReadAndWriteCapacity';
const GSI_TEST_CASE_3 = 'ProjectionTypeKeysOnly';
const GSI_TEST_CASE_4 = 'ProjectionTypeInclude';
const GSI_TEST_CASE_5 = 'InverseTableKeySchema';
const GSI_PARTITION_KEY: Attribute = { name: 'gsiHashKey', type: AttributeType.String };
const GSI_SORT_KEY: Attribute = { name: 'gsiSortKey', type: AttributeType.Number };
const GSI_NON_KEY: string[] = [];
for (let i = 0; i < 20; i++) { // 'A' to 'T'
  GSI_NON_KEY.push(String.fromCharCode(65 + i));
}

const app = new App(process.argv);

const stack = new Stack(app, STACK_NAME);

const table = new Table(stack, 'Table', {
  pitrEnabled: true,
  sseEnabled: true,
  streamSpecification: StreamViewType.KeysOnly,
  ttlAttributeName: 'timeToLive'
});

table.addPartitionKey(TABLE_PARTITION_KEY);
table.addSortKey(TABLE_SORT_KEY);
table.addGlobalSecondaryIndex({
  indexName: GSI_TEST_CASE_1,
  partitionKey: GSI_PARTITION_KEY,
});
table.addGlobalSecondaryIndex({
  indexName: GSI_TEST_CASE_2,
  partitionKey: GSI_PARTITION_KEY,
  sortKey: GSI_SORT_KEY,
  readCapacity: 10,
  writeCapacity: 10,
});
table.addGlobalSecondaryIndex({
  indexName: GSI_TEST_CASE_3,
  partitionKey: GSI_PARTITION_KEY,
  sortKey: GSI_SORT_KEY,
  projectionType: ProjectionType.KeysOnly,
});
table.addGlobalSecondaryIndex({
  indexName: GSI_TEST_CASE_4,
  partitionKey: GSI_PARTITION_KEY,
  sortKey: GSI_SORT_KEY,
  projectionType: ProjectionType.Include,
  nonKeyAttributes: GSI_NON_KEY
});
table.addGlobalSecondaryIndex({
  indexName: GSI_TEST_CASE_5,
  partitionKey: TABLE_SORT_KEY,
  sortKey: TABLE_PARTITION_KEY,
});

const tableWithoutSecondaryIndex = new Table(stack, 'TableWithoutSecondaryIndex', {});
tableWithoutSecondaryIndex.addPartitionKey(TABLE_PARTITION_KEY);

process.stdout.write(app.run());
