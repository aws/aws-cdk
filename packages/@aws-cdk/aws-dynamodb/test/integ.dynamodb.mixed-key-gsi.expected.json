{
  "Resources": {
    "TableCD117FA1": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "KeySchema": [
          {
            "AttributeName": "pkey",
            "KeyType": "HASH"
          }
        ],
        "AttributeDefinitions": [
          {
            "AttributeName": "pkey",
            "AttributeType": "N"
          },
          {
            "AttributeName": "foo",
            "AttributeType": "S"
          },
          {
            "AttributeName": "baz",
            "AttributeType": "S"
          },
          {
            "AttributeName": "bar",
            "AttributeType": "S"
          }
        ],
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "IndexA",
            "KeySchema": [
              {
                "AttributeName": "foo",
                "KeyType": "HASH"
              }
            ],
            "Projection": {
              "NonKeyAttributes": [
                "bar"
              ],
              "ProjectionType": "INCLUDE"
            },
            "ProvisionedThroughput": {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5
            }
          },
          {
            "IndexName": "IndexB",
            "KeySchema": [
              {
                "AttributeName": "baz",
                "KeyType": "HASH"
              },
              {
                "AttributeName": "bar",
                "KeyType": "RANGE"
              }
            ],
            "Projection": {
              "NonKeyAttributes": [
                "blah"
              ],
              "ProjectionType": "INCLUDE"
            },
            "ProvisionedThroughput": {
              "ReadCapacityUnits": 5,
              "WriteCapacityUnits": 5
            }
          }
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    }
  }
}